core/src/forwarding_stage.rs

```diff
@@ -0,0 +1,616 @@
+            root_bank_cache,
+            forward_address_getter,
+            connection_cache,
+            vote_client: VoteClient::new(),
+            data_budget,
+            metrics: ForwardingStageMetrics::default(),
+        }
+    }
+
+    /// Runs `ForwardingStage`'s main loop, to receive, order, and forward packets.
+    fn run(mut self) {
+        loop {
+            let root_bank = self.root_bank_cache.root_bank();
+            if !self.receive_and_buffer(&root_bank) {
+                break;
+            }
+            self.forward_buffered_packets();
+            self.metrics.maybe_report();
+        }
+    }
+
+    /// Receive packets from previous stage and insert them into the buffer.
+    fn receive_and_buffer(&mut self, bank: &Bank) -> bool {
+        // Timeout is long enough to receive packets but not too long that we
+        // forward infrequently.
+        const TIMEOUT: Duration = Duration::from_millis(10);
+
+        let now = Instant::now();
+        match self.receiver.recv_timeout(TIMEOUT) {
+            Ok((packet_batches, tpu_vote_batch)) => {
+                self.metrics.did_something = true;
+                self.buffer_packet_batches(packet_batches, tpu_vote_batch, bank);
+
+                // Drain the channel up to timeout
+                let timed_out = loop {
+                    if now.elapsed() >= TIMEOUT {
+                        break true;
+                    }
+                    match self.receiver.try_recv() {
+                        Ok((packet_batches, tpu_vote_batch)) => {
+                            self.buffer_packet_batches(packet_batches, tpu_vote_batch, bank)
+                        }
+                        Err(_) => break false,
+                    }
+                };
+
+                // If timeout waas reached, prevent backup by draining all
+                // packets in the channel.
+                if timed_out {
+                    warn!("ForwardingStage is backed up, dropping packets");
+                    while let Ok((packet_batch, _)) = self.receiver.try_recv() {
+                        self.metrics.dropped_on_timeout +=
+                            packet_batch.iter().map(|b| b.len()).sum::<usize>();
+                    }
+                }
+
+                true
+            }
+            Err(RecvTimeoutError::Timeout) => true,
+            Err(RecvTimeoutError::Disconnected) => false,
+        }
+    }
+
+    /// Insert received packets into the packet container.
+    fn buffer_packet_batches(
+        &mut self,
+        packet_batches: BankingPacketBatch,
+        is_tpu_vote_batch: bool,
+        bank: &Bank,
+    ) {
+        for batch in packet_batches.iter() {
+            for packet in batch
+                .iter()
+                .filter(|p| initial_packet_meta_filter(p.meta()))
+            {
+                let Some(packet_data) = packet.data(..) else {
+                    // should never occur since we've already checked the
```

Comment:
if it should never occur, shouldn't we assert?

Reply:
Should never happen, but if a packet does have no data for some reason (malformed meta?) I'd rather we just skip that packet than crash the validator for what is very much a secondary service.

Reply:
should count occurrences in metrics

Reply:
#4991